functorch/_src/eager_transforms.py

```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
         grad, _ = results
         return grad
     return wrapper
+
+
+def assert_wrapped_functional(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _assert_wrapped_functional(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def propagate_functional_input_mutations(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _propagate_functional_input_mutation(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def functionalize(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        try:
+            func_level = _func_increment_nesting()
+            func_args = [_wrap_functional_tensor(x, func_level) if isinstance(x, torch.Tensor) else x for x in args]
```

Comment:
We could just tree_map _wrap_functional_tensor on the Tensors to support cases where the user passes in e.g. dicts of tensors. Otherwise, it might be a good idea to assert that the input is a flat list of tensors.

Reply:
Yep thanks - tree_map() seems like the right thing here (I'll probably take a look at how it's used in the other transforms)